roomserver/storage/shared/storage.go

```diff
@@ -343,14 +343,52 @@ func (d *Database) GetMembership(ctx context.Context, roomNID types.RoomNID, req
 }
 
 func (d *Database) GetMembershipEventNIDsForRoom(
-	ctx context.Context, roomNID types.RoomNID, joinOnly bool, localOnly bool,
+	ctx context.Context, roomNID types.RoomNID, joinOnly bool, membershipFilter []string, localOnly bool,
 ) ([]types.EventNID, error) {
 	if joinOnly {
 		return d.MembershipTable.SelectMembershipsFromRoomAndMembership(
 			ctx, roomNID, tables.MembershipStateJoin, localOnly,
 		)
 	}
 
+	if len(membershipFilter) > 0 {
+		var eventNIDs []types.EventNID
+		var filteredEventNIDs []types.EventNID
+		var err error
+		leaveOrBan := false
+		for _, filter := range membershipFilter {
+			if filter == "join" {
+				filteredEventNIDs, err = d.MembershipTable.SelectMembershipsFromRoomAndMembership(
+					ctx, roomNID, tables.MembershipStateJoin, localOnly,
+				)
+				if err != nil {
+					return eventNIDs, err
+				}
+				eventNIDs = append(eventNIDs, filteredEventNIDs...)
+			}
+			if filter == "invite" {
+				filteredEventNIDs, err = d.MembershipTable.SelectMembershipsFromRoomAndMembership(
+					ctx, roomNID, tables.MembershipStateInvite, localOnly,
+				)
+				if err != nil {
+					return eventNIDs, err
+				}
+				eventNIDs = append(eventNIDs, filteredEventNIDs...)
```

Comment:
This seems needlessly complex. Ultimately, you're mapping `filter` into a `tables.MembershipStateXXX`. Why can't you just do:
```go
state := tables.MembershipStateLeaveOrBan
switch (filter) {
    case "invite":
        state = tables.MembershipStateInvite
    case "join":
        state = tables.MembershipStateJoin
    case "leave":
        fallthrough
    case "ban":
        state = tables.MembershipStateLeaveOrBan
    default:
        return nil, fmt.Errorf("unknown membership filter: %v", filter)
}

filteredEventNIDs, err = d.MembershipTable.SelectMembershipsFromRoomAndMembership(
    ctx, roomNID, state, localOnly,
)
```